src/control/lib/daos/api/attribute.go

```diff
@@ -0,0 +1,268 @@
+	// before fetching the actual values.
+	cAttrSizes := make([]C.size_t, numAttr)
+	var rc C.int
+	switch at {
+	case poolAttr:
+		rc = daos_pool_get_attr(hdl, C.int(numAttr), &cAttrNames[0], nil, &cAttrSizes[0], nil)
+	/*case contAttr:
+	rc = daos_cont_get_attr(hdl, C.int(numAttr), &attrNames[0], nil, &attrSizes[0], nil)*/
+	default:
+		return nil, errors.Wrapf(daos.InvalidInput, "unknown attr type %d", at)
+	}
+	if err := daosError(rc); err != nil {
+		return nil, errors.Wrapf(err, "failed to get %s attribute sizes", at)
+	}
+
+	// Now, create a slice of buffers to hold the values.
+	cAttrValues := make([]unsafe.Pointer, numAttr)
+	defer func(valueSlice []unsafe.Pointer) {
+		for _, value := range valueSlice {
+			C.free(value)
+		}
+	}(cAttrValues)
+	for i, size := range cAttrSizes {
+		if size < 1 {
+			return nil, errors.Wrapf(daos.MiscError, "failed to get %s attribute %s: size is %d", at, reqAttrNames[i], size)
+		}
+
+		cAttrValues[i] = C.malloc(size)
+	}
+
+	// Do the actual fetch of all values in one go.
+	switch at {
+	case poolAttr:
+		rc = daos_pool_get_attr(hdl, C.int(numAttr), &cAttrNames[0], &cAttrValues[0], &cAttrSizes[0], nil)
+	/*case contAttr:
+	rc = daos_cont_get_attr(hdl, C.int(numAttr), &attrNames[0], &attrValues[0], &attrSizes[0], nil)*/
+	default:
+		return nil, errors.Wrapf(daos.InvalidInput, "unknown attr type %d", at)
+	}
+	if err := daosError(rc); err != nil {
+		return nil, errors.Wrapf(err, "failed to get %s attribute values", at)
+	}
+
+	// Finally, create a slice of attribute structs to hold the results.
+	// Note that we are copying the values into Go-managed byte slices
+	// for safety and simplicity so that we can free the C memory as soon
+	// as this function exits.
+	attrs := make([]*daos.Attribute, numAttr)
+	for i, name := range reqAttrNames {
+		attrs[i] = &daos.Attribute{
+			Name:  name,
+			Value: C.GoBytes(cAttrValues[i], C.int(cAttrSizes[i])),
+		}
+	}
+
+	return attrs, nil
+}
+
+// setDaosAttributes sets the values for the given list of attribute names.
+// Uses the bulk attribute set API to minimize roundtrips.
+func setDaosAttributes(hdl C.daos_handle_t, at attrType, attrs daos.AttributeList) error {
+	if len(attrs) == 0 {
+		return errors.Wrapf(daos.InvalidInput, "no %s attributes provided", at)
+	}
+
+	// First, build a slice of C strings for the attribute names.
+	attrNames := make([]*C.char, len(attrs))
+	for i, attr := range attrs {
+		if attr == nil {
+			return errors.Wrapf(daos.InvalidInput, "nil %s attribute at index %d", at, i)
+		}
+		if attr.Name == "" {
+			return errors.Wrapf(daos.InvalidInput, "empty %s attribute name at index %d", at, i)
+		}
+		attrNames[i] = C.CString(attr.Name)
+	}
+	defer func(nameSlice []*C.char) {
+		for _, name := range nameSlice {
+			freeString(name)
+		}
+	}(attrNames)
+
+	// Next, create a slice of C.size_t entries to hold the sizes of the values,
+	// and a slice of pointers to the actual values.
```

Comment:
Any reason this step couldn't be done at the same time as the previous one? I understand why we couldn't do that for the getter, but the setter should be able to build all of these arrays at the same time.

Reply:
This code has been working as-is for a couple of years now. My inclination is to get it all moved inside the API with proper unit tests first, and then work on cleaning up inefficiencies with confidence that the unit tests will find any mistakes.

Reply:
Okay, fair point.